### ISSUE TITLE
Pulling updated widget values from front-end

### ISSUE THREAD

**Issue by hannorein** [jupyter-widgets/ipywidgets#1287]
# Summary:

Add the examples in https://github.com/jupyter-widgets/ipywidgets/issues/1287#issuecomment-301982976 and https://github.com/jupyter-widgets/ipywidgets/issues/1287#issuecomment-301990689 to the docs as an advanced example for having a request/reply pattern with widgets using generator-like syntax

# Original issue

I can check for updates to a widget value using `observe`:

```python
import ipywidgets as widgets
from IPython.display import display
import time

w = widgets.Dropdown(
    options=['Addition', 'Multiplication', 'Subtraction'],
    value='Addition',
    description='Task:',
)

def on_change(change):
    print("changed to %s" % change['new'])

w.observe(on_change)

display(w)
```

But how can I do that during another operation, for example:

```python
for i in range(100):
    time.sleep(1)
    # pull for changes to w here.
    # if w.has_changed:
    #     print(w.value)
```

I tried doing something like 

```python
import IPython
ipython = IPython.get_ipython()
ipython.kernel.do_one_iteration()
```
to pull for updates. But is seems unreliable. Is there a better way to do this?

**Comment by jasongrout** [comment]
We don't have a good way to poll for updates. Fundamentally, we're working on top of the kernel messaging protocol, and updates are messages that have to be processed like any other message, which in Python can't happen while the current code is running. Furthermore, the problem is fundamentally asynchronous, so it calls for a different approach. That said, you may have some luck doing an async/await construct to structure your code in an async way.

**Comment by hannorein** [comment]
Thanks. I tried to play around with aync/wait but no luck so far. Is there any other way I can send data from javascript to python and wait for it to arrive? It doesn't seem like such a crazy workflow. Maybe I could hijack the web server and have my own web socket...

**Comment by jasongrout** [comment]
> Is there any other way I can send data from javascript to python and wait for it to arrive?

There isn't infrastructure to deal with having python code pause execution and wait for events from javascript (except for the raw_input/input python builtin functions to get raw strings from the front end). Options for dealing with this include: (a) using a callback-based system, i.e., register a callback to continue your code and end your current execution, or (b) start another thread with your computation and return from the main thread and use some sort of locking primitive and a callback handler on the main thread to unblock the other thread, or (c) do some async/await magic to make the callback-based system in (a) look like synchronous code (I think this would work...).

CCing @minrk in case he has ideas too.

Can you elaborate on what you are trying to do? There may be a more natural way to accomplish your end goal.

**Comment by jasongrout** [comment]
(labeling as no action for now since we're still in a discussion about how to do things)

**Comment by hannorein** [comment]
@jasongrout In short: I have an interactive widget that visualizes an N-body simulation with WebGL. I'd like to allow the user to take a screenshot of the WebGL canvas while the simulation is running. For that, I need to send the data back from JS to python, save it to a file and only then continue running the simulation. 

I now found a workaround using the `observe` callback that works (more or less, the syntax is a bit ugly). Basically what I'm doing is integrating the simulation to the first time where I want a screenshot, then setting up a callback function so I get notified when the screenshot data arrives. And then I'm restarting the simulation in the callback function until the next output.

**Comment by jasongrout** [comment]
In bqplot, we do this same thing completely client-side. There is a button near the rendered plot, and when the user clicks the button, it downloads the png through their web browser, rather than transferring it back to the kernel. The kernel knows nothing about the screenshot.

That means it is impossible to programmatically control the screenshot. If you need that, your way of doing it is the callback-oriented approach, which is probably the cleanest way right now.

CC @dmadeka, @ssunkara1.

**Comment by hannorein** [comment]
Ok. Thanks. Yes, I have to do it programmatically if I want to capture thousands of frames to render a movie. (the rendering, transfer and saving of images is really fast, by the way).

**Comment by jasongrout** [comment]
It would be interesting to see if we could come up with an async/await solution. Also, I think @takluyver was working on an asynchronous kernel - that could be useful here too.

**Comment by takluyver** [comment]
Might be @Carreau that you're thinking of.

**Comment by jasongrout** [comment]
> Might be @Carreau that you're thinking of.

Ah yes, thanks!

**Comment by jasongrout** [comment]
The async kernel mentioned above is here: https://github.com/ipython/ipython/pull/10390

**Comment by Carreau** [comment]
Yes, and it's far from being a full async kernel. That would be need major rewrite of IPython and lot's of API changes. 

Maybe once we drop 3.4 :-)

**Comment by jasongrout** [comment]
Closing (but labeled as a reference issue) as a workaround was found, and doing something more would likely require fundamental changes to the messaging protocol or implementation and an issue in that repo.

**Comment by jasongrout** [comment]
It might be interesting to write a decorator that essentially turns an iterator into a request/reply system on widgets. For example, something like:

```python
@pause(widget, 'take_screenshot', 'screenshot')
def run_computation():
    for i in range(10):
        print(i)
        # do some computation
        screenshot = yield i
        print(screenshot)
run_computation()
```
which would register an observe handler for the widget. Every time the iterator yields, it sets the widget 'take_screenshot' value to i. When the frontend sees the value of 'take_screenshot' change, it takes a screenshot, which syncs the data back to the kernel. The observe handler for the screenshot attribute then calls the iterator `send()` method with the screenshot value, and you continue the function.

**Comment by hannorein** [comment]
That's a good idea. Thanks for the tips!

**Comment by jasongrout** [comment]
Here's a working implementation:

```python
from functools import wraps
def pause(send, receive):
    def f(iterator):
        @wraps(iterator)
        def inner():
            i = iterator()
            def next_i(change):
                try:
                    send[0].set_trait(send[1], i.send(change.new))
                except StopIteration as e:
                    send[0].set_trait(send[1], e.value)
                    receive[0].unobserve(next_i, receive[1])
            receive[0].observe(next_i, receive[1])
            send[0].set_trait(send[1], next(i))
        return inner
    return f
```
```python
from ipywidgets import IntSlider, VBox, HTML
a=HTML(description='set by function:')
b=IntSlider(description='send to function:')
VBox([a,b])
```
```python
@pause((a,'value'), (b,'value'))
def f():
    for i in range(10):
        # do some work
        bvalue = yield 'Did work %s'%i
        print('Received value %s'%bvalue)
    return 'Done!'
f()
```

**Comment by jasongrout** [comment]
so in this example, the function runs until the yield, and then sets the a.value trait to what is yielded and pauses. When the b.value trait changes (like the user clicks on a slider, or in our case maybe the screenshot is automatically triggered by setting a.value), the function resumes, and keeps doing so until the function ends and sets the a.value to 'Done' and exits (which unregisters the observe handler).

Basically, this is just converting the iterator code to the callback approach...

Even better might be to have async/await syntax, but that I think requires tapping into the event loop, or maybe starting up another event loop.

**Comment by jasongrout** [comment]
Reopening so that we can put this example in the docs as an advanced use case for using widgets and needing to have a conversation back and forth with widgets, but wanting to write the conversation in a single function.

**Comment by jasongrout** [comment]
Here is an alternative implementation using async/await:

(first cell due to https://github.com/ipython/ipykernel/issues/21#issue-82784960)
```python
import asyncio
from ipykernel.eventloops import register_integration

@register_integration('asyncio')
def loop_asyncio(kernel):
    '''Start a kernel with asyncio event loop support.'''
    loop = asyncio.get_event_loop()

    def kernel_handler():
        loop.call_soon(kernel.do_one_iteration)
        loop.call_later(kernel._poll_interval, kernel_handler)

    loop.call_soon(kernel_handler)
    try:
        if not loop.is_running():
            loop.run_forever()
    finally:
        loop.run_until_complete(loop.shutdown_asyncgens())
        loop.close()
```
```python
%gui asyncio
```
```python
import asyncio
def wait_for_change(widget, value):
    future = asyncio.Future()
    def getvalue(change):
        future.set_result(change.new)
        widget.unobserve(getvalue, value)
    widget.observe(getvalue, value)
    return future
```
```python
from ipywidgets import IntSlider
a = IntSlider()
a
```
```python
async def f():
    for i in range(10):
        print('did work %s'%i)
        x = await wait_for_change(a, 'value')
        print('async function continued with value %s'%x)
asyncio.ensure_future(f())
```

**Comment by jasongrout** [comment]
I like the async/await version better - it more clearly says just "pause and wait for this widget value to change"

**Comment by jasongrout** [comment]
(see also https://github.com/ipython/ipykernel/pull/246, where I put in a PR to ipykernel to add asyncio integration support to the %gui magic)

**Comment by jasongrout** [comment]
@hannorein - #1361 contains some nice solutions than the ones above (in particular, a nicer generator solution).

**Comment by hannorein** [comment]
@jasongrout Thanks! I'll look into it (I'll probably only get to it after the weekend).
